packages/simulator_2d/src/simulator_engine.cpp

```diff
@@ -0,0 +1,150 @@
+    double integration_step, double precision): model_(model_config_path) {
+    
+    params_.integration_step = integration_step;
+    params_.precision = precision;
+
+    cache_.integration_step_2 = integration_step / 2;
+    cache_.integration_step_6 = integration_step / 6;
+    cache_.inverse_integration_step = 1 / integration_step;
+    cache_.inverse_wheelbase_length = 1 / model_.wheelBase().length;
+    cache_.wheelbase_width_2 = model_.wheelBase().width / 2;
+
+    reset();
+}
+
+void SimulatorEngine::reset(double x, double y, double yaw, double steering, 
+        double linear_velocity, double angular_velocity) {
+    state_ = (SimulatorEngine::State() 
+        << x, y, yaw, steering, linear_velocity, angular_velocity)
+        .finished();
+}
+
+void SimulatorEngine::reset() {
+    reset(-model_.wheelBase().base_to_rear, 0, 0, 0, 0, 0);
+}
+
+rclcpp::Time SimulatorEngine::getTime() const {
+    return time_;
+}
+
+geom::Pose SimulatorEngine::getPose() const {
+    geom::Pose pose;
+    pose.pos.x = state_[StateIndex::x] + model_.wheelBase().base_to_rear;
+    return pose;
+}
+
+double SimulatorEngine::getCurrentCurvature() const {
+    return model_.wheelBase().length * tan(state_[StateIndex::steering]);
+}
+
+double SimulatorEngine::getMiddleSteering() const {
+    return state_[StateIndex::steering];
+}
+
+model::Steering SimulatorEngine::getCurrentSteering() const {
+    return model_.rearCurvatureToSteering(getCurrentCurvature());
+}
+
+model::Steering SimulatorEngine::getTargetSteering() const {
+    return model_.rearCurvatureToSteering(control_.curvature);
+}
+
+model::Twist SimulatorEngine::getTwist() const { 
+    const auto twist = model::Twist {getCurrentCurvature(), state_[StateIndex::linear_velocity]};
+    return model_.rearToBaseTwist(twist);
+}
+
+geom::Vec2 SimulatorEngine::getLinearVelocity() const {
+    return geom::Vec2::fromAngle(geom::Angle(getTwist().velocity));
+}
+
+geom::Vec2 SimulatorEngine::getAngularVelocity() const {
+    return geom::Vec2::fromAngle(geom::Angle(state_[StateIndex::angular_velocity]));
+}
+
+void SimulatorEngine::setControl(
+    double velocity, double acceleration, double curvature) {
+
+    control_.velocity 
+        = model_.baseToLimitedRearVelocity(velocity, curvature);
+    control_.acceleration 
+        = model_.baseToLimitedRearAcceleration(acceleration, curvature);
+    control_.curvature 
+        = model_.baseToLimitedRearCurvature(curvature);
+}
+
+void SimulatorEngine::setControl(double velocity, double curvature) {
+    control_.velocity 
+        = model_.baseToLimitedRearVelocity(velocity, curvature);
+    control_.curvature 
+        = model_.baseToLimitedRearCurvature(curvature);
+
+    const auto limits = model_.baseAccelerationLimits();
+    if (state_[StateIndex::linear_velocity] - params_.precision < control_.velocity) {
+        control_.acceleration = limits.max;
+    } else {
+        control_.acceleration = limits.min;
+    }

```

Comment:
Почему так? А если назад поедем?
`acc_max` – ускорение набор скорости
`acc_min` – ускорение торможения

Reply:
Написал сейчас вот так:
```suggestion
    const auto limits = model_.baseAccelerationLimits();
    if (rear_ax_state_[StateIndex::linear_velocity] * velocity > 0) {
        const double current_speed 
            = abs(rear_ax_state_[StateIndex::linear_velocity] - params_.precision);
        if (current_speed < abs(velocity)) {
            acceleration = limits.max;
        } else {
            acceleration = limits.min;
        }

        if (velocity < 0) {
            acceleration *= -1;
        }
    }
    else {
        acceleration = limits.min;

        if (velocity > 0) {
            acceleration *= -1;
        }
    }

    setBaseControl(velocity, acceleration, curvature);
```

1. Если текущая скорость и требуемая одного знака, то мы просто разгоняемся или тормозим.
2. Иначе нам в любом случае нужно сначала остановиться, а потом уже разгоняться в другую сторону.
3. Для разгона в другую сторону в методе advance при прохождении через 0 проверяем, тормозили мы сейчас или нет. Если да, то переключаем флаг в "разгон" и берём значение ускорения из model.

Очень интересно ваше мнение.

Reply:
Возможно, флаг не нужен и достаточно в моменте сравнивать текущую и требуемую скорости.